modules/hitlnext/src/backend/service.ts

```diff
@@ -0,0 +1,112 @@
+import * as sdk from 'botpress/sdk'
+import _ from 'lodash'
+
+import { Config } from '../config'
+import { MODULE_NAME } from '../constants'
+import { ExitTypes, IHandoff, ISocketMessage } from '../types'
+import { StateType } from '.'
+import Repository from './repository'
+
+export const toEventDestination = (
+  botId: string,
+  handoff: Pick<IHandoff, 'userId' | 'userThreadId' | 'userChannel'>
+): sdk.IO.EventDestination => {
+  return { botId, target: handoff.userId, threadId: handoff.userThreadId, channel: handoff.userChannel }
+}
+
+interface Realtime {
+  sendPayload: (botId: string, message: ISocketMessage) => void
+}
+
+class Service {
+  constructor(
+    private bp: typeof sdk,
+    private state: StateType,
+    private repository: Repository,
+    private realtime: Realtime
+  ) {}
+
+  async createHandoff(
+    botId: string,
+    dest: Pick<IHandoff, 'userId' | 'userThreadId' | 'userChannel' | 'status'>,
+    timeoutDelay: number
+  ) {
+    const config: Config = await this.bp.config.getModuleConfigForBot(MODULE_NAME, botId)
+    const eventDestination = toEventDestination(botId, dest)
+    const attributes = await this.bp.users.getAttributes(dest.userChannel, dest.userId)
+    const language = attributes.language
+
+    const handoff = await this.repository.createHandoff(botId, dest).then(handoff => {
+      this.state.cacheHandoff(botId, handoff.userThreadId, handoff)
+      return handoff
+    })
+
+    if (config.transferMessage) {
+      await this.sendMessageToUser(config.transferMessage, eventDestination, language)
+    }
+
+    this.realtime.sendPayload(botId, { resource: 'handoff', type: 'create', id: handoff.id, payload: handoff })
+
+    if (timeoutDelay !== undefined && timeoutDelay > 0) {
+      setTimeout(async () => {
+        const userHandoff = await this.repository.getHandoff(handoff.id)
+
+        if (userHandoff.status === 'pending') {
+          await this.updateHandoff(userHandoff, botId, { status: 'expired' })
+          await this.transferToBot(eventDestination, 'timedOutWaitingAgent')
+        }
+      }, timeoutDelay * 1000)
```

Comment:
PP: You should think about clearing the timeout when the handoff is assigned so that we don't have to make a DB call for nothing.